### DISCUSSION TITLE
Using files from the "source" in "output"

### DISCUSSION THREAD

**Discussion by gothy** [vercel/community#61] [Help] [answered]
Hi there

I have a need to provide my backend code with `GOOGLE_APPLICATION_CREDENTIALS` env var that should point to the file containing firebase app creds.
I have that file in the root of the sources dir. Listing this file with `./` prefix path in the Now env vars doesn't make much difference, i think it's because the output is missing most of the files from the source. 
Is there a way to specify which files i want to move to "output" during the build process?

**Comment by lucleray** [comment, marked as answer]
Hi @gothy ,

You can use custom [Build Command](https://vercel.com/docs/v2/build-step#build-command) and [Output Directory](https://vercel.com/docs/v2/build-step#output-directory) to add custom logic to your Build and copy source files over to the output directory.

For example, if you are using Next.js, you can set your Build Command to `next build && cp my-firebase-creds.txt output/my-firebase-creds.txt` and your Output Directory to `output`.

Note that this will publish your file and **your file will be exposed to your visitors**.

If you want to keep your file content secret, you will need to find a way to pass the content of the file instead of a file path to the firebase library you are using.

**Reply by gothy** [reply]
> Note that this will publish your file and your file will be exposed to your visitors.

Wow, that's a serious limitation to some apps, i guess.

> If you want to keep your file content secret, you will need to find a way to pass the content of the file instead of a file path to the firebase library you are using.

Yep, I have found a workaround for this specific case, but there might be others.

@lucleray thanks for your help!

**Reply by rafinskipg** [reply]
hello @gothy @lucleray could you provide a more detailed example on how to use the GOOGLE_APPLICATION_CREDENTIALS in vercel? 

I need the credentials file to be able to use google APIs but I don't know how to upload it

**Reply by gothy** [reply]
@rafinskipg for my particular issue, i had to initialize firebase app with google creds.
I've ended up using `credential.cert` instead of `credential.applicationDefault` and `require`ing the json with admin key directly:

```js
  fbadmin.initializeApp({
    // credential: fbadmin.credential.applicationDefault(),
    credential: fbadmin.credential.cert(require('../../firebaseAdminKey.json')),
    databaseURL: 'https://my-project-name.firebaseio.com',
  });
```

Hope this helps :)

**Reply by rafinskipg** [reply]
hello @gothy  thanks for sharing. This means that you uploaded the firebaseAdminKey into the project repository?

**Reply by gothy** [reply]
@rafinskipg yes for now.

**Reply by rafinskipg** [reply]
@lucleray As gothy says, it is a serious limitation that deployments on vercel can not be configured to use google application credentials. Is there any RFC for that? 
I am sure the purpose of vercel is to provide a number 1 developer experience, and this could be an important point. 

The support of Vercel sent me this idea also https://vercel.com/paulo/encrypted-data/mucp1kd62/source, by encripting and de-encrypting the data. Which i don't like also

**Reply by lucleray** [reply]
@rafinskipg Looking at the [Firebase docs](https://firebase.google.com/docs/reference/admin/node/admin.credential#cert), you can pass the value of the credentials (instead of the path) to initialize the SDK.

It looks like this with the code @gothy shared previously:
```js
  fbadmin.initializeApp({
    credential: fbadmin.credential.cert(JSON.parse(process.env.FIREBASE_ADMIN_KEY)),
    databaseURL: 'https://my-project-name.firebaseio.com',
  });
```

You can use [`now env add`](https://vercel.com/docs/now-cli#commands/env/extended-usage) to upload the content of a file as an Environment Variable like this:
```
vercel env add FIREBASE_ADMIN_KEY preview < firebaseAdminKey.json
```
(Replace `preview` with `production` if you want to provide the secret for your Production Deployments)

That way, you don't have to upload it in your repo, and you can rely on our system encrypting and storing safely your secrets.

Does this work for your use case?

**Reply by rafinskipg** [reply]
I am not using firebase, but normal google cloud. Google cloud authentication 

In my use case I am using Google Cloud storage (https://www.npmjs.com/package/@google-cloud/storage) 

Diving into the code i can see that there are another options available, like using `keyFileName`, but it also requires a File, and that means having it copied to your vercel project.

````javascript
 new Storage({ projectId:
   'your-project-id', keyFilename: '/path/to/keyfile.json'
 });
````
Another option would be to use "credentials" and passing a "client_email" and a "private_key". This would be awesome if it worked.  I also tried passing the whole configuration object here, but nothing.

I used my client_email and private_key from my service account, but it does not work. See [here](https://github.com/googleapis/nodejs-storage/blob/master/src/storage.ts#L359) and here
https://googleapis.dev/nodejs/storage/latest/global.html#StorageOptions


@lucleray , what i also tried todo is doing the following

1) Override build command in Vercel, and make a `git clone` from a repository with my credentials
2) Try to set the environment variable pointing to that repository files. But it does not work, maybe the absolute path i am using is wrong. It is not clearly explained in the documentation which is the absolute path of the deployment. I have tried this `GOOGLE_APPLICATION_CREDENTIALS=~/mycloned-repo/file.json` and also `./mycloned-repo/file.json`. Both failing



UPDATE: 

This may fail in the future, but at the end it worked by *generating a new service account file* and introducing it into the `credentials` options in the storage constructor. But i fear that this method will only work for 60 minutes. The previous file i was using it worked fine when set as GOOGLE_APPLICATION_CREDENTIALS but failed when used as a `credentials` property, saying that the token was short lived and already expired.
My guess is that setting GOOGLE_APPLCIATION_CREDENTIALS will do a refresh token, but using credentials will not.

Lets see. I will come back to update

**Reply by rafinskipg** [reply]
The app it's been running for a couple of hours with no problem.

Final solution: 

```
const storage = new Storage({
  credentials: {
    client_email: process.env.GOOGLE_CLIENT_EMAIL,
    private_key: process.env.GOOGLE_PRIVATE_KEY.replace(/\\n/g, '\n')
  },
  projectId: process.env.GOOGLE_PROJECTID
})
````

**Reply by cdock1029** [reply]
@rafinskipg Following this blog post, here's possibly a cleaner way to do it without worrying about newlines and such. You can base64 encode/decode the file from a secret

https://dev.to/parondeau/gcp-credentials-next-js-3a0d

**Reply by rafinskipg** [reply]
Wow @cdock1029  thanks for sharing! I will keep it in my list of resources.

**Reply by saitodisse** [reply]
> @rafinskipg Following this blog post, here's possibly a cleaner way to do it without worrying about newlines and such. You can base64 encode/decode the file from a secret
> 
> [dev.to/parondeau/gcp-credentials-next-js-3a0d](https://dev.to/parondeau/gcp-credentials-next-js-3a0d)

this is the best way. Use JSON.parse to get the privatekey: https://github.com/vercel/vercel/issues/749#issuecomment-715009494
